### ISSUE TITLE
 Inconsistent local download archive path  - causing issue with local_path_to_archive

### ISSUE THREAD

**Issue by iamgini** [ansible-middleware/wildfly#249]
##### SUMMARY

Noticed the controller and EE getting different paths when running from controller and running locally using automation controller (the dev and testing stage)

##### ISSUE TYPE
 - Bug Report

##### ANSIBLE VERSION

```
$ ansible-navigator exec -- ansible --version
ansible [core 2.15.3]
  config file = /home/iamgini/ansible-bau/ansible-jboss-automation/ansible.cfg
  configured module search path = ['/home/runner/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /home/iamgini/ansible-bau/ansible-jboss-automation/collections:/home/runner/.ansible/collections/ansible_collections:/usr/share/ansible/collections/ansible_collections
  executable location = /usr/bin/ansible
  python version = 3.9.16 (main, May 31 2023, 12:21:58) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (/usr/bin/python3.9)
  jinja version = 3.1.2
  libyaml = True
```

`ansible-navigator.yml`

```yaml
---
ansible-navigator:
  execution-environment:
    container-engine: podman
    image: registry.redhat.io/ansible-automation-platform-24/ee-supported-rhel8:1.0.0-395
    # image: localhost/ee-supported-rhel8-expect:1.0
    enabled: true
    pull:
      policy: missing
    # volume-mounts:
    #   - src: "/home/iamgini/projects/Scripts"
    #     dest: "/opt/apps"
    #     options: "Z"
  editor:
    command: code-server {filename}
    console: false

  logging:
    # level: debug
    append: false

  mode: stdout

  playbook-artifact:
    enable: false
    replay: artifacts/ansible_artifact.json
    save-as: artifacts/ansible_artifact.json
```

##### COLLECTION VERSION

```
$ ansible-galaxy collection list | grep eap
redhat.eap                                    1.4.3  
```

##### STEPS TO REPRODUCE

```yaml
---
- name: Ensure eap is install and running as a service
  hosts: all
  vars_files:
    - vars/vars.yml
  vars:
    eap_offline_install: true
    eap_version: 8.0.0
    eap_apply_cp: false
  roles:
    - eap_install
    - eap_systemd
  tasks:
    - name: Install second driver with eap_driver role
      ansible.builtin.include_role:
        name: eap_driver
      when: jdbc_drivers is defined and jdbc_drivers | length > 0
      vars:
        eap_driver_module_name: "{{ item.name }}"
        eap_driver_version: "{{ item.version }}"
        eap_driver_jar_filename: "{{ item.jar_file }}"
        eap_driver_jar_url: "{{ item.url }}"
      loop: "{{ jdbc_drivers }}"
```

##### EXPECTED RESULTS

The `local_path.stat.path` should be consistent when we run it from the controller and ansible-navigator. 

##### ACTUAL RESULTS

```
$ ansible-navigator run jboss_eap_deployment.yaml -e "nodes=jboss"  -vvv
```

The suspected task:

in `roles/eap_install/tasks/install.yml`

```yaml
- name: Check local download archive path
  ansible.builtin.stat:
    path: "{{ lookup('env', 'PWD') }}"
  register: local_path
  delegate_to: localhost
```

Running from automation controller

```json 
{
  "ansible_facts": {
    "full_path_to_archive": "/opt/jboss_eap//jboss-eap-8.0.0.zip",
    "local_path_to_archive": "/runner/project/jboss-eap-8.0.0.zip"
  },
  "_ansible_no_log": null,
  "changed": false
}
```

Running local using ansible-navigatror

```json
ok: [rhel8-jboss] => {
    "ansible_facts": {
        "full_path_to_archive": "/opt/jboss_eap//jboss-eap-8.0.0.zip",
        "local_path_to_archive": "/home/iamgini/ansible-bau/ansible-jboss-automation/jboss-eap-8.0.0.zip"
    },
    "changed": false
}
```

**Comment by RanabirChakraborty** [comment]
Hello, @iamgini looks like you are using the old eap collection, Can you please use the latest one?
in the latest collection, we have introduced a new variable `wildfly_install_download_controller_download_dir` which can probably help you in this scenario. 

```
- name: Check local download archive path
  ansible.builtin.stat:
    path: "{{ wildfly_install_download_controller_download_dir | default(lookup('env', 'PWD')) }}"
  register: local_path
  delegate_to: localhost
  ```
 
Checkout the PR as well - https://github.com/ansible-middleware/wildfly/pull/216

**Comment by iamgini** [comment]
@RanabirChakraborty oh ok. Thanks for letting me know.
Actually, the customer is using an old EE and might be using the old collection inside. 
I will try this and update the issue.

**Comment by rpelisse** [comment]
@sabre1041 @guidograzioli do we provided our collection within an execution environment? If so, we probably need to update this one.

**Comment by iamgini** [comment]
I can see the following check fails.

The task should check `wildfly_install_download_controller_download_dir/eap_archive_filename_local` right?

```yaml
- name: "Validate existing zipfiles {{ eap_archive_filename }} for offline installs"
  ansible.builtin.assert:
    that:
      - eap_version is defined and eap_archive_filename_local is exists
    quiet: True
    fail_msg: "An offline install was requested, but files are not present on the controller: {{ eap_archive_filename }}"
  when:
    - eap_offline_install
```

**Comment by rpelisse** [comment]
@iamgini That really points toward the fact that the zipfile is NOT on the **controller** (or not at the right place).

I recall from our discussion elsewhere, that you mentioned you could NOT put the zipfile on the controller (so I suspect you put it on the target). If so, we need to change this check because the following stmt **only** works on the controller:

```... eap_archive_filename_local is exists
```

Can you simply remove this part of the code in our local testing env (just edit the file in ~/.ansible/collections/...) and test, to confirm this is the culprit?

**Comment by rpelisse** [comment]
About 2) : The idea is that  customers would implement this kind of things (like you did) in the `pre_tasks:` of the playbook, so that zipfile is in place when the collection. I don't think we should add this sort of commidity to the collection, because, in essence, we would just be wrapping `ansible.builtin.get_url:`  and having to deal with ALL the options (like auth) coming with it.

However, we do need to fix that the collection **assumes** the zipfile will be on the controller. It's better and it makes sens (avoid download the zipfile on EACH target is smoother for performance), but it should be possible for user to use zipfile already available on targets (for instance, if the zipfile is mounted from a remote file system on the target).

**Comment by iamgini** [comment]
@rpelisse you are right; I agree on item 2.

Yeah, for offline files, maybe some logic to tell the collection that, "file is on controller" or "file is on target node already". And yes, a full/absolute path validation for "exist" check 😉 

Thanks

**Comment by rpelisse** [comment]
@iamgini Sorry to come back to you on this so late, but I'm finally working on a fix and I have a question about this request from you:
> And yes, a full/absolute path validation for "exist" check 😉

Actually, you can the var `wildfly_archive_filename_local` to whatever you want (and thus an absolute path, if you need to), so I'm not sure if we need to really change the check.

Other than that, I'm working on skipping this test if the zipfile is, indeed, on the target.

**Comment by rpelisse** [comment]
@iamgini I'm closing the issue, as I've fixed the problem with the check. if you still want to discuss the absolute path thing you wanted to see implement, just open a new issue!

**Comment by rpelisse** [comment]
Fixed by [#249](https://github.com/ansible-middleware/wildfly/issues/249)
